substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/AbstractGarbageCollectorMXBean.java

```diff
@@ -0,0 +1,134 @@
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.core.genscavenge;
+
+import com.oracle.svm.core.util.BasedOnJDKFile;
+
+import com.oracle.svm.core.SubstrateUtil;
+import com.oracle.svm.core.genscavenge.service.GcNotificationRequest;
+import com.oracle.svm.core.genscavenge.service.PoolMemoryUsage;
+import com.oracle.svm.core.genscavenge.service.Target_com_sun_management_GcInfo;
+import com.oracle.svm.core.genscavenge.service.Target_com_sun_management_internal_GcInfoBuilder;
+
+import com.sun.management.GarbageCollectionNotificationInfo;
+import com.sun.management.GcInfo;
+import com.sun.management.internal.GarbageCollectionNotifInfoCompositeData;
+import com.sun.management.internal.GcInfoBuilder;
+import sun.management.NotificationEmitterSupport;
+
+import javax.management.MBeanNotificationInfo;
+import javax.management.Notification;
+import javax.management.NotificationEmitter;
+import javax.management.openmbean.CompositeData;
+import java.lang.management.MemoryUsage;
+
+public abstract class AbstractGarbageCollectorMXBean extends NotificationEmitterSupport
+                implements com.sun.management.GarbageCollectorMXBean, NotificationEmitter {
+
+    @BasedOnJDKFile("https://github.com/openjdk/jdk/blob/jdk-24+14/src/hotspot/share/gc/serial/serialHeap.cpp#L451") //
+    private static final String ACTION_MINOR = "end of minor GC";
+
+    @BasedOnJDKFile("https://github.com/openjdk/jdk/blob/jdk-24+14/src/hotspot/share/gc/serial/serialHeap.cpp#L718") //
+    private static final String ACTION_MAJOR = "end of major GC";
+
+    private GcInfoBuilder gcInfoBuilder;
+    private volatile GcInfo gcInfo;
+    private long seqNumber = 0;
+
+    private synchronized GcInfoBuilder getGcInfoBuilder() {
+        if (gcInfoBuilder == null) {
+            Target_com_sun_management_internal_GcInfoBuilder gib = new Target_com_sun_management_internal_GcInfoBuilder(this, getMemoryPoolNames());
+            gcInfoBuilder = SubstrateUtil.cast(gib, GcInfoBuilder.class);
+        }
+        return gcInfoBuilder;
+    }
+
+    /**
+     * Use the data taken from the request queue to populate MemoryUsage. The service thread calls
+     * this method.
+     */
+    public void createNotification(GcNotificationRequest request) {
```

Comment:
Parts of this code are probably based on OpenJDK code. If so, please annotate with `@BasedOnJDKFile`.